### PR TITLE
feature/player-interaction

### PR DESCRIPTION
**Pull request by Joost-Robat**

### CODE REVIEW

File: Assets/Scripts/Framework/Gameplay/Player/Interactor.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+        private void OnEnable() => _inputParser.OnInteract += UseInteractable;
+
+        private void OnDisable() => _inputParser.OnInteract -= UseInteractable;
+
+        private void OnTriggerEnter(Collider collider)
+        {
+            GameObject target = collider.gameObject;
+
+            if (target.TryGetComponent<Interactable>(out Interactable interactable))
```
Comment (PiterGroot) [inline Assets/Scripts/Framework/Gameplay/Player/Interactor.cs]:
```suggestion
            if (target.TryGetComponent(out Interactable interactable))
```
This is redundant, type can already be retrieved by out keyword

File: Assets/Scripts/Framework/Gameplay/Player/Interactor.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+                _interactables.Add(interactable);
+            }
+        }
+
+        private void OnTriggerExit(Collider collider)
+        {
+            GameObject target = collider.gameObject;
+
+            if(target.TryGetComponent<Interactable>(out Interactable interactable))
```
Comment (PiterGroot) [inline Assets/Scripts/Framework/Gameplay/Player/Interactor.cs]: Same thing

**Review by PiterGroot — CHANGES_REQUESTED** [review]
Nice stuff

File: Assets/Scripts/Framework/Gameplay/Interactions/Interactable.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        public Action OnInteract;
+
+        [SerializeField]
+        private float _cooldown = .5f;
+
+        private protected bool IsOnCooldown { get; set; }
+
+        private protected bool IsInteractable { get; set; } = true;
+
+        private protected bool IsBeingInteracted { get; set; }
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Interactions/Interactable.cs]: `IsBeingInteracted` is only used and set in `Grabbable` so I think it is better to move it to that script.

File: Assets/Scripts/Framework/Gameplay/Interactions/Interactable.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+        /// <summary>
+        /// A function that calls a RPC to the server on this Interactable.
+        /// </summary>
+        /// <param name="interactor">The interactor that his request originates from.</param>
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Interactions/Interactable.cs]:
```suggestion
        /// <param name="interactor">The interactor that this request originates from.</param>
```

File: Assets/Scripts/Framework/Gameplay/Player/Interactor.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                return;
+
+            if (_isInteracting)
+            {
+                StartCoroutine(ActivateCooldown());
+                _interactingTarget.InteractOnServer(this);
+                return;
+            }
+
+            // TO DO: Priority checking for which interactable is most prio.
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Player/Interactor.cs]:
```suggestion
```

File: Assets/Scripts/Framework/Gameplay/Player/Interactor.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            _interactingTarget = interactable;
+            _isInteracting = isInteracting;
+
+            if (_isInteracting)
+                interactable.NetworkObject.SetParent(_grabbingAnchor.GetComponent<NetworkBehaviour>());
+            else
+                interactable.NetworkObject.UnsetParent();
+
+            if(_isInteracting)
+                interactable.gameObject.transform.localPosition = Vector3.zero;
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Player/Interactor.cs]: You can remove this `if` statement and move the line of code after the `SetParent()` line inside the `if` statement.

File: Assets/Scripts/Framework/Gameplay/Player/PlayerMovement.cs [hunk 7]
```diff
@@ -46,14 +48,25 @@ private void OnDisable()
...
 
             newForce.x = Mathf.Clamp(newForce.x, _maxSpeed * -1, _maxSpeed);
             newForce.z = Mathf.Clamp(newForce.z, _maxSpeed * -1, _maxSpeed);
 
             _rigidbody.velocity = newForce;
+
+            if (_playerInput.normalized != Vector2.zero)
+            {
+                gameObject.transform.rotation =
+                    Quaternion.Lerp
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Player/PlayerMovement.cs]: Maybe we can try using `Slerp` instead of `Lerp`.

**Review by lucasvdiepen — CHANGES_REQUESTED** [review]

**Review by PiterGroot — APPROVED** [review]

File: Assets/Scripts/Framework/Gameplay/Player/PlayerMovement.cs [hunk 8]
```diff
@@ -46,14 +48,25 @@ private void OnDisable()
...
+            if (_playerInput.normalized != Vector2.zero)
+            {
+                gameObject.transform.rotation =
+                    Quaternion.Slerp
+                    (
+                        gameObject.transform.rotation,
+                        Quaternion.LookRotation(playerInput),
+                        Mathf.Clamp01(_rotationSpeed * Time.deltaTime)
+                    );
+            }
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Gameplay/Player/PlayerMovement.cs]: Make a new method for this rotation functionality.

**Review by lucasvdiepen — CHANGES_REQUESTED** [review]

**Review by Joost-Robat — COMMENTED** [review]
Changes of @lucasvdiepen looking good! ;)

**Review by lucasvdiepen — APPROVED** [review]

**Review by PiterGroot — APPROVED** [review]
LGTM

**Pull request merged**